### ISSUE TITLE
root path exists is inconsistent between osfs and s3. 

### ISSUE THREAD

**Issue by simonm3** [PyFilesystem/pyfilesystem2#303]
Sometimes I want to know if a folder exists. Would be useful if this returned true or false:

> f = fs.open_fs("filesystem"://path/to/folder")
> f.exists("")

On S3 this returns True even though file does not exist.

> f = fs.open_fs("s3://registr1/thisfiledoesnotexist")
> f.exists("")

On OSFS this throws a "CreateFailed" exception

> f = fs.open_fs("osfs://c:/thisfiledoesnotexist")
> f.exists("")

**Comment by simonm3** [comment]
Related to this I want to split an fs_url into filesystem/path. Would be useful to have a function to do this that works on all filesystems. 

The rules vary by filesystem. On S3 I need to include the bucket name in the filesystem; and on windows I need to include the drive in the filesystem. Maybe on other filesystems there are different rules? I am using the code below for s3/osfs.

```
# default filesystem
        if "://" not in url:
            url = f"osfs://{url}"

        # parse url. note parse_result.path appears to be unused
        parse_result = parse_fs_url(url)
        protocol  = parse_result.protocol.lower()
        self.path = parse_result.resource
        self.root = f"{self.protocol}://"

        # s3 root must include bucket
        if protocol=="s3":
            parts = fs.path.parts(self.path)
            self.root.append(parts[0])
            self.path = parts[1:]

        # windows root must include drive
        elif protocol=="osfs":
            res = self.path.split(":/")
            if len(res)==1:
                pass
            elif len(res)==2:
                self.root.append(f"{res[0]}:/")
                self.path = res[1]
            else:
                raise Exception("invalid path")
```

**Comment by lurch** [comment]
I _suspect_ what you're trying to do in your first example is probably filesystem-specific, and therefore "undefined behaviour"?

Why not adapt your examples to:
```python
f = fs.open_fs("s3://registr1/")
f.exists("thisfiledoesnotexist")
```
and
```python
f = fs.open_fs("osfs://c:/")
f.exists("thisfiledoesnotexist")
```
?

@willmcgugan Is it (or should it be) "valid" for `.exists` to be given an empty path?

**Comment by simonm3** [comment]
I am using pyfilesystem to avoid anything filesystem specific. I could
adapt my examples as you suggest but then I am introducing filesystem
specific code to split the bucket for s3; or the drive for windows; and
likely in other ways for other file systems. Of course I will workaround
but would be great if pyfilesystem handled all the unique behaviour in one
way.

**Comment by lurch** [comment]
@simonm3 I'm afraid I don't understand your response. IMHO wanting to do `f.exists("")` seem like a very strange thing to do.
_Perhaps_ what you want is an optional argument to the `open_fs` function (which I guess would then be handled by each of the FS classes' `__init__` method) that checks that the supplied "opener path" (i.e. "s3://registr1/thisfiledoesnotexist" and "osfs://c:/thisfiledoesnotexist" in your earlier examples) actually accesses a valid filesystem?

**Comment by simonm3** [comment]
Agree exists(" ") should raise an exception but whatever it does it should
be consistent.

What I want to do is treat an fs_url as a filepath that is agnostic to the
file system. I have a wrapper that does this already but has to treat each
file system differently. This would be better included in pyfs. This would
include calling "exists" but also other functions such as "open", "save",
"load", "isdir". Currently the fs_url is in two parts - filesystem and path
which are file system specific. The bucket on s3 and drive on windows is
part of the filesystem.

**Comment by lurch** [comment]
@simonm3 Maybe it'd help if you tried to explain what it is you're doing?

I think the general pyfilesystem "usage model" is:

1. open fs object with init-args
2. do stuff with paths within the fs object
3. do more stuff with paths within the fs object
4. close the fs object

It's only step 1 in the list above that _isn't_ filesystem-agnostic, and the "init-args" would include the drive-path on Windows (or the initial path on Linux) or the bucket on S3.  (I've never used S3, so apologies if I misuse any terms)

**Comment by willmcgugan** [comment]
I think I see @simonm3 's issue. There is an inconsistency in the API here.

So OSFS won't let you create an instance when the root doesn't exist. That seems reasonable (at the time at least) because the object would be useless if the root path didn't exist. Alas, I probably hadn't considered network filesystems at the time. The trouble with network filesystems is that to implement the same `CreateFailed` behaviour the constructor would have to make an expensive network request, and I would prefer to keep the constructor light-weight. Which is likely why it was left out of S3FS and FTPFS.

The other inconsistency is in `exists("")` or `exist("/")`. I think I may have assumed that the root is always going to exist which isn't always the case (seems obvious now). So in some filesystems it may return True by default, on others it may actually do something.

What to do about that? I'm not certain. But obviously being consistent would be my preferred option.
 
@simonm3 with regards to the FS URLs, they do parse in to common fields, but beyond that it is filesystem dependent.

**Comment by lurch** [comment]
> the constructor would have to make an expensive network request, and I would prefer to keep the constructor light-weight.

That was why I was suggesting an optional `is_fs_valid` (or something) init-parameter, which would obviously be False by default.

@willmcgugan Semantically, what would it mean for a filesystem to be opened, but for it to have a non-existent root? Is the cat dead or alive? :wink:

Off the top of my head, I think it _might_ make sense for the pyfilesystem API to "assume" that the root always exists, for `exists("")` to throw a `InvalidPath` error, and for any operations on (e.g.) a network-FS that's been constructed with a non-existent root to throw a `ResourceNotFound` error. Maybe the behaviour of `exists("/")` still ought to be implementation-dependent? :man_shrugging:  (And maybe providing a `is_fs_valid=True` parameter to the `__init__` method, ought to then add a call to `self.exists("/")` at the end of `__init__`, throwing Exceptions as necessary?)
Perhaps for consistency, **any** pyfilesystem method that's given a path of "" ought to throw a `InvalidPath` error?

@simonm3 just to clarify on my earlier point, the 'problem' with trying to parse a `osfs://` URL into a filesystem-part and a path-part, is that you can't tell where the filesystem-part stops and the path-part starts.
E.g. the following would be equivalent:
```python
f = fs.open_fs("osfs://c:/")
f.exists("thisfolderexists/thisfiledoesnotexist")
```
or
```python
f = fs.open_fs("osfs://c:/thisfolderexists")
f.exists("thisfiledoesnotexist")
```

@willmcgugan Hmm, what happens if `open_fs` is given a path to a file, rather than a path to a directory? Perhaps the [documentation](https://docs.pyfilesystem.org/en/latest/openers.html) needs clarifying here?

**Comment by willmcgugan** [comment]
> That was why I was suggesting an optional is_fs_valid (or something) init-parameter, which would obviously be False by default.

That would solve it. Although that would be a pain to introduce to all filesystems. I think I'd prefer to just say that the constructor should just write attributes, and not attempt to do anything with the underlaying storage - which is desirable for network filesystems.

Come to think of it, the solution may already be in the API. There is a [check](https://docs.pyfilesystem.org/en/latest/reference/base.html#fs.base.FS.check) method which is mainly there to throw an exception if you try to use a closed filesystem, but may be the right place to check the root actually exists. So you might do this:

```
f = fs.open_fs("osfs://c:/thisfiledoesnotexist")
f.check()
```

That feels safer than relying on a side-effect of `exists` and related methods. By making the check explicit, we wouldn't need to ensure that all methods respond the same when there is something fundamentally broken.

Do you not think that `exists("")` and `exists("/")` should be equivalent? Given that the paths are relative to the root `"/foo"` is equivalent to `"foo"`. So shouldn't `"/"` be equivalent to `""` ?

`exists("/")` is an odd one isn't it? It's entirely possible for the root to exist when the filesystem is created, and then not exists later. It could even come back in the lifetime of the filesystem. 

> @willmcgugan Semantically, what would it mean for a filesystem to be opened, but for it to have a non-existent root? Is the cat dead or alive? 😉

If anything has been opened, its a can of worms...

> @willmcgugan Hmm, what happens if open_fs is given a path to a file, rather than a path to a directory? Perhaps the documentation needs clarifying here?

I think that should raise OpenerError, but I suspect it might not. Yeah, the docs should be clearer here.

**Comment by lurch** [comment]
> By making the check explicit, we wouldn't need to ensure that all methods respond the same when there is something fundamentally broken.

Yeah, that might work; but presumably it would throw something other than `FilesystemClosed` if the root directory doesn't exist?

> Do you not think that exists("") and exists("/") should be equivalent?

Personally, no. One is explicitly giving a root directory, and the other is giving what (IMHO) is the equivalent of `None`. :man_shrugging: 
But (again IMHO), given that pyfs doesn't support the concept of a CWD, I've never understood why it had "absolute" paths anyway, rather than always only using relative paths...

> It's entirely possible for the root to exist when the filesystem is created, and then not exists later. It could even come back in the lifetime of the filesystem.

Yeah, that's possible even with OSFS:
```python
f = fs.open_fs("osfs://c:/somedirectory")
# do stuff with f
shutil.rmtree("c:/somedirectory")
# oops, f has no root directory
os.mkdir("c:/somedirectory")
# whoo, f has a root directory again
```

> If anything has been opened, its a can of worms...

Lol, sorry!

> I think that should raise OpenerError, but I suspect it might not.

I wonder if it'd be worth providing a `fs.opener.download_file` method that _could_ be given a fs-URL pointing to a file? (which would error if pointed at a directory)
